File: clients/python/tests/test_send.py
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

"""Producer-side tests for ``Client.send`` / ``Client.send_batch``."""

from unittest.mock import MagicMock, call

import pytest

import pgque


def test_send_returns_int_event_id(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    eid = client.send(queue, {"order_id": 42})
    assert isinstance(eid, int)
    assert eid > 0


def test_send_with_explicit_type(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    eid = client.send(queue, {"id": 1}, type="order.created")
    assert isinstance(eid, int)


def test_send_event_object(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    eid = client.send(queue, pgque.Event(payload={"x": 1}, type="custom.t"))
    assert isinstance(eid, int)


def test_send_str_payload_passes_through(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    eid = client.send(queue, '"plain string"')
    assert isinstance(eid, int)


def test_send_none_payload(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    eid = client.send(queue, None)
    assert isinstance(eid, int)


def test_send_unicode_payload(conn, setup_queue):
    queue, consumer = setup_queue
    client = pgque.PgqueClient(conn)
    payload = {"text": "héllo wörld 🎉 — ünicode тест"}
    client.send(queue, payload)
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    msgs = client.receive(queue, consumer, max_messages=10)
    assert len(msgs) == 1
    got = msgs[0].payload if isinstance(msgs[0].payload, dict) \
        else __import__("json").loads(msgs[0].payload)
    assert got == payload
    client.ack(msgs[0].batch_id)
    conn.commit()


def test_send_large_payload(conn, setup_queue):
    queue, consumer = setup_queue
    client = pgque.PgqueClient(conn)
    big = {"data": "x" * 100_000}
    client.send(queue, big)
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    msgs = client.receive(queue, consumer, max_messages=10)
    assert len(msgs) == 1
    got = msgs[0].payload if isinstance(msgs[0].payload, dict) \
        else __import__("json").loads(msgs[0].payload)
    assert got == big
    client.ack(msgs[0].batch_id)
    conn.commit()


def test_send_batch_returns_ids_in_order(conn, setup_queue):
    queue, _ = setup_queue
    client = pgque.PgqueClient(conn)
    ids = client.send_batch(queue, "batch.test", [
        {"n": 1}, {"n": 2}, {"n": 3}, {"n": 4},
    ])
    assert len(ids) == 4
    assert all(isinstance(i, int) for i in ids)
    assert ids == sorted(ids)


def test_send_to_missing_queue_raises(conn):
    client = pgque.PgqueClient(conn)
    with pytest.raises(pgque.PgqueError):
        client.send("does_not_exist_xyz_12345", {"x": 1})
        conn.commit()
    conn.rollback()


@pytest.mark.parametrize("type_val,expect_3arg", [
    (None, False),       # None → 2-arg (default type)
    ("", False),         # "" → 2-arg (empty = default)
    ("default", False),  # "default" → 2-arg
    ("custom", True),    # non-default → 3-arg
])
def test_send_sql_form_selection(type_val, expect_3arg):
    """send() picks 2-arg SQL when type is empty/None/default; 3-arg otherwise."""
    mock_cursor = MagicMock()
    mock_cursor.fetchone.return_value = (999,)
    mock_conn = MagicMock()
    mock_conn.execute.return_value = mock_cursor

    client = pgque.PgqueClient(mock_conn)
    if type_val is None:
        eid = client.send("q", {"x": 1}, type=None)
    else:
        eid = client.send("q", {"x": 1}, type=type_val)

    assert eid == 999
    sql_used = mock_conn.execute.call_args[0][0]
    if expect_3arg:
        assert "send(%s, %s, %s::jsonb)" in sql_used, (
            f"expected 3-arg form for type={type_val!r}, got: {sql_used!r}"
        )
    else:
        assert "send(%s, %s::jsonb)" in sql_used, (
            f"expected 2-arg form for type={type_val!r}, got: {sql_used!r}"
        )
        assert "send(%s, %s, %s::jsonb)" not in sql_used


@pytest.mark.parametrize("payload,expected", [
    ({"key": "val", "n": 1}, {"key": "val", "n": 1}),  # dict
    ([1, "two", None], [1, "two", None]),               # list
    ('"just a string"', "just a string"),               # JSON string literal
    ("42", 42),                                         # JSON number
    ("null", None),                                     # JSON null
])
def test_jsonb_payload_round_trip(conn, setup_queue, payload, expected):
    """jsonb payloads decode to native Python types after send/receive."""
    import json

    queue, consumer = setup_queue
    client = pgque.PgqueClient(conn)
    client.send(queue, payload)
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    msgs = client.receive(queue, consumer, max_messages=10)
    assert len(msgs) == 1
    raw = msgs[0].payload
    got = raw if not isinstance(raw, str) else json.loads(raw)
    assert got == expected
    client.ack(msgs[0].batch_id)
    conn.commit()


def test_send_batch_mixed_payloads_preserve_order(conn, setup_queue):
    """send_batch preserves payload order, including JSON null."""
    import json

    queue, consumer = setup_queue
    client = pgque.PgqueClient(conn)
    payloads = [{"a": 1}, None, "42"]
    expected = [{"a": 1}, None, 42]
    ids = client.send_batch(queue, "batch.mixed", payloads)
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    msgs = client.receive(queue, consumer, max_messages=10)
    assert [m.msg_id for m in msgs] == ids
    got = [
        m.payload if not isinstance(m.payload, str) else json.loads(m.payload)
        for m in msgs
    ]
    assert got == expected
    client.ack(msgs[0].batch_id)
    conn.commit()


def test_send_batch_none_payload_produces_json_null(conn, setup_queue):
    """send_batch([None]) must store JSON null, not SQL NULL."""
    # send(None) coerces to JSON null via "null"; send_batch must match it.
    # Passing Python None through psycopg as SQL NULL would bypass ::jsonb.
    import json

    queue, consumer = setup_queue
    client = pgque.PgqueClient(conn)
    client.send_batch(queue, "default", [None])
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    msgs = client.receive(queue, consumer, max_messages=10)
    assert len(msgs) == 1, "expected exactly 1 message from send_batch([None])"
    raw = msgs[0].payload
    # payload must be JSON null (Python None after psycopg decode), not a
    # missing/SQL-NULL value that would cause a TypeError on json.dumps.
    got = raw if not isinstance(raw, str) else json.loads(raw)
    assert got is None, (
        f"send_batch([None]) should store JSON null; got {raw!r}"
    )
    client.ack(msgs[0].batch_id)
    conn.commit()

File: clients/ruby/test/test_send.rb
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

require_relative "test_helper"

class TestSend < Minitest::Test
  include PgqueTest::Helpers

  def test_send_returns_int_event_id
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      eid = client.send(queue, { "order_id" => 42 })
      assert_kind_of Integer, eid
      assert_operator eid, :>, 0
    end
  end

  def test_send_with_explicit_type
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      eid = client.send(queue, { "id" => 1 }, type: "order.created")
      assert_kind_of Integer, eid
    end
  end

  def test_send_event_object
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      event = Pgque::Event.new(payload: { "x" => 1 }, type: "custom.t")
      eid = client.send(queue, event)
      assert_kind_of Integer, eid
    end
  end

  def test_send_str_payload_passes_through
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      eid = client.send(queue, '"plain string"')
      assert_kind_of Integer, eid
    end
  end

  def test_send_nil_payload
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      eid = client.send(queue, nil)
      assert_kind_of Integer, eid
    end
  end

  def test_send_numeric_and_boolean_payloads_coerce_via_to_s
    # Non-String/Hash/Array/nil payloads run through to_s so numerics
    # and booleans round-trip naturally as JSON scalars.
    cases = [
      [42,    42],
      [3.14,  3.14],
      [true,  true],
      [false, false],
    ]
    cases.each do |payload, expected|
      with_queue do |queue, consumer, conn|
        client = Pgque::Client.new(conn)
        client.send(queue, payload)
        conn.exec_params("select pgque.force_next_tick($1)", [queue])
        conn.exec_params("select pgque.ticker($1)", [queue])
        msgs = client.receive(queue, consumer, 10)
        assert_equal 1, msgs.size, "no message for #{payload.inspect}"
        assert_equal expected, msgs[0].payload,
                     "#{payload.inspect} did not round-trip"
        client.ack(msgs[0].batch_id)
      end
    end
  end

  def test_send_batch_returns_ids_in_order
    with_queue do |queue, _consumer, conn|
      client = Pgque::Client.new(conn)
      ids = client.send_batch(queue, "batch.test", [
        { "n" => 1 }, { "n" => 2 }, { "n" => 3 }, { "n" => 4 }
      ])
      assert_equal 4, ids.size
      assert ids.all? { |i| i.is_a?(Integer) }
      assert_equal ids.sort, ids
    end
  end

  def test_send_unicode_payload
    with_queue do |queue, consumer, conn|
      client = Pgque::Client.new(conn)
      payload = { "text" => "héllo wörld 🎉 — ünicode тест" }
      client.send(queue, payload)
      conn.exec_params("select pgque.force_next_tick($1)", [queue])
      conn.exec_params("select pgque.ticker($1)", [queue])
      msgs = client.receive(queue, consumer, 10)
      assert_equal 1, msgs.size
      assert_equal payload, msgs[0].payload
      client.ack(msgs[0].batch_id)
    end
  end

  def test_send_large_payload
    with_queue do |queue, consumer, conn|
      client = Pgque::Client.new(conn)
      big = { "data" => "x" * 100_000 }
      client.send(queue, big)
      conn.exec_params("select pgque.force_next_tick($1)", [queue])
      conn.exec_params("select pgque.ticker($1)", [queue])
      msgs = client.receive(queue, consumer, 10)
      assert_equal 1, msgs.size
      assert_equal big, msgs[0].payload
      client.ack(msgs[0].batch_id)
    end
  end

  def test_jsonb_payload_round_trip
    cases = [
      [{ "key" => "val", "n" => 1 }, { "key" => "val", "n" => 1 }],
      [[1, "two", nil],              [1, "two", nil]],
      ['"just a string"',             "just a string"],
      ["42",                          42],
      ["null",                        nil],
    ]
    cases.each do |payload, expected|
      with_queue do |queue, consumer, conn|
        client = Pgque::Client.new(conn)
        client.send(queue, payload)
        conn.exec_params("select pgque.force_next_tick($1)", [queue])
        conn.exec_params("select pgque.ticker($1)", [queue])
        msgs = client.receive(queue, consumer, 10)
        assert_equal 1, msgs.size, "no message for payload=#{payload.inspect}"
        if expected.nil?
          assert_nil msgs[0].payload,
                     "payload=#{payload.inspect} did not round-trip to nil"
        else
          assert_equal expected, msgs[0].payload,
                       "payload=#{payload.inspect} did not round-trip"
        end
        client.ack(msgs[0].batch_id)
      end
    end
  end

  def test_send_batch_mixed_payloads_preserve_order
    with_queue do |queue, consumer, conn|
      client = Pgque::Client.new(conn)
      payloads = [{ "a" => 1 }, nil, "42"]
      expected = [{ "a" => 1 }, nil, 42]
      ids = client.send_batch(queue, "batch.mixed", payloads)
      conn.exec_params("select pgque.force_next_tick($1)", [queue])
      conn.exec_params("select pgque.ticker($1)", [queue])
      msgs = client.receive(queue, consumer, 10)
      assert_equal ids, msgs.map(&:msg_id)
      assert_equal expected, msgs.map(&:payload)
      client.ack(msgs[0].batch_id)
    end
  end

  def test_send_batch_nil_payload_produces_json_null
    with_queue do |queue, consumer, conn|
      client = Pgque::Client.new(conn)
      client.send_batch(queue, "default", [nil])
      conn.exec_params("select pgque.force_next_tick($1)", [queue])
      conn.exec_params("select pgque.ticker($1)", [queue])
      msgs = client.receive(queue, consumer, 10)
      assert_equal 1, msgs.size, "send_batch([nil]) should produce 1 message"
      assert_nil msgs[0].payload, "payload must be JSON null, not SQL NULL"
      client.ack(msgs[0].batch_id)
    end
  end

  def test_send_to_missing_queue_raises
    conn = PG.connect(PGQUE_TEST_DSN)
    begin
      client = Pgque::Client.new(conn)
      assert_raises(Pgque::Error) do
        client.send("does_not_exist_xyz_12345", { "x" => 1 })
      end
    ensure
      conn.close
    end
  end
end

class TestSendSqlForm < Minitest::Test
  # Capture exec_params calls without a real DB.
  class FakeConn
    attr_reader :sql_used, :params_used

    def exec_params(sql, params)
      @sql_used = sql
      @params_used = params
      FakeResult.new
    end

    class FakeResult
      def getvalue(_row, _col)
        "999"
      end
    end
  end

  def test_2arg_form_for_default_type
    [nil, "", "default"].each do |type_val|
      conn = FakeConn.new
      client = Pgque::Client.new(conn)
      eid = client.send("q", { "x" => 1 }, type: type_val)
      assert_equal 999, eid
      assert_includes conn.sql_used, "send($1, $2::jsonb)"
      refute_includes conn.sql_used, "send($1, $2, $3::jsonb)",
                      "type=#{type_val.inspect} should use 2-arg form"
    end
  end

  def test_3arg_form_for_custom_type
    conn = FakeConn.new
    client = Pgque::Client.new(conn)
    eid = client.send("q", { "x" => 1 }, type: "custom")
    assert_equal 999, eid
    assert_includes conn.sql_used, "send($1, $2, $3::jsonb)"
  end
end
